### ISSUE TITLE
Small bounding boxes not producing any loss

### ISSUE THREAD

**Issue by ervteng** [balancap/SSD-Tensorflow#222]
Hi all, I noticed an issue with the loss computation for small (less than 1/10 of the image dimensions) bounding boxes. 

It seems that given such a bounding box as ground truth, the loss function will always output 0 loss. I believe the issue is that the ground truth bounding box is smaller than any of the anchor boxes, and does not overlap suitably (IOU of 0.5) with any of them - thus, no match. This means that small object augmentation is basically useless, as the loss will always be 0 and no learning happens. Is this a limitation of MultiBox in general or is particular to this repo? 

I am aware of SSD's known deficiencies with small objects, but from what I can gather it is expected to be a graceful degradation (i.e., lower and lower confidence). Rather, in this case, it seems that small bounding boxes are ignored entirely during training.

Thanks!

**Comment by ervteng** [comment]
I confirmed that this is because of the IOU/jaccard overlap threshold, as reducing this threshold does produce a reasonable loss output for very small bounding boxes. 

I'm not sure what the original paper does. The paper reads:

> We begin by matching each ground truth box to the default box with the **best
jaccard overlap (as in MultiBox)**. Unlike MultiBox, we then match default boxes to
any ground truth with jaccard overlap higher than a threshold (0.5).

This could be interpreted as (best + >0.5), but most implementations interpret this as filtering out all matches below 0.5, including the best overlap. In my tests, keeping the best overlap in addition to others with jaccard >0.5 seems to work fine, and helps with training on very small bounding boxes. I don't believe this is an issue on the PASCAL VOC dataset, as none of those bounding boxes are that small.

**Comment by bainjamain** [comment]
I think that what the paper means is that every ground truth is matched **at least** with one bounding box, i.e. the one with the greatest Jaccard overlap (even if this overlap is below .5).

**Comment by ervteng** [comment]
As the current implementation does _not_ do this (it could be 0 matches), I was able to do that by adding this line in ssd_losses, after `pmask = gscores >matching_threshold`.

```
pmask = tf.concat(axis=0, values=[pmask[:tf.argmax(gscores)], [True], pmask[tf.argmax(gscores)+1:]])
```

**Comment by bainjamain** [comment]
Thanks @ervteng for sharing!

**Comment by bainjamain** [comment]
@ervteng your snippet should be adapted to handle multiple objects as well as batch > 1. For the batch a simple for loop could work, not very elegant but that would make the job.

**Comment by ervteng** [comment]
@benmyara Good call! I got multiple batch sizes to work - had to rewrite everything between `fglocalizations = []` and `fpmask = tf.cast(pmask, dtype)`, as the way the original code concats the vectors makes it very difficult to extract on a per-batch level:

```
# First concat all the vectors batch-wise
        for i in range(len(logits)):
            flogits.append(tf.reshape(logits[i], [batch_size, -1, num_classes]))
            fgclasses.append(tf.reshape(gclasses[i], [batch_size, -1]))
            fgscores.append(tf.reshape(gscores[i], [batch_size, -1]))
            flocalisations.append(tf.reshape(localisations[i], [batch_size, -1, 4]))
            fglocalisations.append(tf.reshape(glocalisations[i], [batch_size, -1, 4]))

        logits = tf.concat(flogits, axis=1)
        gclasses = tf.concat(fgclasses, axis=1)
        gscores = tf.concat(fgscores, axis=1)
        localisations = tf.concat(flocalisations, axis=1)
        glocalisations = tf.concat(fglocalisations, axis=1)

        # We want to keep each of the top scores within each batch. Do this by setting the
        # max score to the match threshold
        _gscores_arr = []
        for j in range(batch_size):
            _gscores = gscores[j]
            max_idx = tf.cast(tf.argmax(_gscores),tf.int32) # neccessary for 32-bit GPUs
            _gscores_arr.append(tf.concat(axis=0, values=[_gscores[:max_idx], [match_threshold], _gscores[max_idx+1:]]))

        gscores = tf.stack(_gscores_arr)
        # Now flatten the vectors
        logits = tf.reshape(logits, [-1, num_classes])
        gclasses = tf.reshape(gclasses, [-1])
        gscores = tf.reshape(gscores, [-1])
        localisations = tf.reshape(localisations, [-1,4])
        glocalisations = tf.reshape(glocalisations, [-1, 4])

        dtype = logits.dtype

        # Compute positive matching mask... but always choose max val
        #max_idx = tf.cast(tf.argmax(gscores),tf.int32)
        #print(gscores)
        pmask = gscores >= match_threshold
``` 

Now, I'm still not sure how to do multiple objects (luckily my application only needs 1 object) but hopefully we can figure it out!

**Comment by bainjamain** [comment]
@ervteng Nice, but you also assume that an image always contains a box. If an image does not contain a bounding box, you will still get a `match`. You should add a condition like: 

```
cond = tf.greater(tf.reduce_sum(_gscores), 0.0001) # if there is a ground truth box
_gscores_modified = tf.concat(axis=0, values=[_gscores[:max_idx], [match_threshold], _gscores[max_idx+1:]])
_gscores = tf.cond(cond, lambda: _g_scores_modified, lambda: _gscores)
_gscores_arr.append(_gscores)
```

**Comment by davinca** [comment]
Thanks for your explanation @ervteng

**Comment by ervteng** [comment]
Thanks @benmyara for that clarification. A quick question for anyone who's more familiar with this codebase that I am - how do you introduce a no object ground truth label into the system? The loss function takes in classification, localizations, and scores, which are outputted by the `bboxes_encode` method.

The `bboxes_encode` method takes two tensors, one a matrix of bounding boxes and one a matrix of class labels. Would you give `bboxes_encode` two empty tensors, or None as the elements?

**Comment by bainjamain** [comment]
@ervteng One way to go about it would be to give `tf_ssd_bboxes_encode_layer` a tensor with one box with only zeros since it would result in feature_scores being only zeros, i.e. it will end up in the negative mask.

**Comment by ervteng** [comment]
Thanks @benmyara. Would that mean labels = [0], bbox = [0,0,0,0]? I found that bbox= [0,0,1,1] technically works, but that seems incorrect.

**Comment by bainjamain** [comment]
@ervteng Yes, with a bounding box equal to `[0., 0., 0., 0.]`, `jaccard_with_anchors` would get `h=w=0.` and would return `0.` which is what we want.
If you chose your box to be `[0., 0., 1., 1.]`, you will have `h=ymax-ymin`, `w=xmax-xmin` and it would return something different from `0.`. I don't see why in this case that would work.

**Comment by ervteng** [comment]
@benmyara I think your approach is working. One small change needs to be made to the hard negative mining mechanism. `nmask = tf.logical_and(nmask, nvalues < max_hard_pred)` should be changed to `nmask = tf.logical_and(nmask, nvalues <= max_hard_pred)`. It seems the original code always selects n_neg -1 negatives, which means that when there is no positive and n_neg is 1, you get 0 negatives in the negative mask. 

I don't believe there is a "canonical" answer to what to do in this case as the original paper does not address the case where there is no positive (PASCAL VOC has no such images). But this approach does "work" in a sense!

**Comment by lyltencent** [comment]
Hi @ervteng , @benmyara , I have a similar problem of applying SSD for small object detection and have tried out the above code. However, it still didn't work for me. I am just wondering if I have changed other part of the code (e.g. anchor_sizes, and anchor_ratios) to get it working?

Thanks in advance!

**Comment by lyltencent** [comment]
Attach the loss during training process: 
![screen shot 2018-11-20 at 4 56 16 pm](https://user-images.githubusercontent.com/10439779/48812019-3b382000-ece5-11e8-8bf0-489f21328302.png)
Thanks!

**Comment by JiangniHIT** [comment]
@lyltencent How could you get that result? My loss is just around 5\6
